### PR TITLE
fix: normalize text-only content arrays to strings for OpenAI-compatible providers

### LINKED ISSUE #50107: [Bug]: NVIDIA Provider Sends Anthropic-Style Message Format to OpenAI-Compatible API
### Bug type

Behavior bug (incorrect output/state without crash)

### Summary

When spawning a subagent with `sessions_spawn` using an NVIDIA provider configured as `"api": "openai-completions"`, OpenClaw sends Anthropic-style message content (arrays with `type` objects) instead of OpenAI-style simple strings. This causes the NVIDIA API to reject requests with HTTP 400 errors.

bug was created using openclaw *

## Configuration

```json
{
  "models": {
    "providers": {
      "nvidia": {
        "baseUrl": "https://integrate.api.nvidia.com/v1",
        "apiKey": "nvapi-XXXXXXXXXX",
        "api": "openai-completions",
        "models": [
          {
            "id": "qwen/qwen2.5-7b-instruct",
            "name": "Qwen 2.5 7B Instruct",
            "reasoning": false,
            "input": ["text"],
            "cost": {
              "input": 0,
              "output": 0,
              "cacheRead": 0,
              "cacheWrite": 0
            },
            "contextWindow": 128000,
            "maxTokens": 8192
          }
        ]
      }
    }
  }
}


## Steps to Reproduce

1. Configure NVIDIA provider with `"api": "openai-completions"`
2. Spawn a subagent using the NVIDIA model:
   ```javascript
   sessions_spawn({
     agentId: "test-agent",
     mode: "run",
     model: "nvidia/qwen/qwen2.5-7b-instruct",
     task: "Say hello",
     runTimeoutSeconds: 60
   })
   ```
3. Observe HTTP 400 error in logs

### Expected behavior

OpenClaw should send OpenAI-compatible message format:

```json
{
  "model": "qwen/qwen2.5-7b-instruct",
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": "Say hello"
    }
  ]
}
```

## Actual Behavior

OpenClaw sends Anthropic-style message format:

```json
{
  "model": "qwen/qwen2.5-7b-instruct",
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": [
        {
          "type": "text",
          "text": "Say hello"
        }
      ]
    }
  ]
}

### OpenClaw version

2026.3.13 (61d171a)

### Operating system

 macOS (Darwin 25.3.0, arm64)

### Install method

npm global

### Model

nvidia/qwen/qwen2.5-7b-instruct

### Provider / routing chain

openclaw -> nvidia

### Config file / key location

~/.openclaw/openclaw.json ; models.providers.nvidia.baseUrl ; https://integrate.api.nvidia.com/v1

### Additional provider/model setup details

(Note: Default is anthropic/claude-sonnet-4-5, but bug occurs when spawning with NVIDIA model)

### Logs, screenshots, and evidence

```shell
Error Logs


2026-03-19T00:20:40.070Z warn agent/embedded {"subsystem":"agent/embedded"} 
{"event":"embedded_run_agent_end","tags":["error_handling","lifecycle","agent_end","assistant_error"],
"runId":"XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX","isError":true,
"error":"HTTP 400: \"[{'type': 'string_type', 'loc': ('body', 'messages', 1, 'content'), 
'msg': 'Input should be a valid string', 'input': [{'type': 'text', 'text': '[Wed 2026-03-18 19:20 CDT] 
[Subagent Context] You are running as a subagent (depth 1/1). Results auto-announce to your requester...",
"failoverReason":"auth","model":"qwen/qwen2.5-7b-instruct","provider":"nvidia",
"rawErrorPreview":"400 \"[{'type': 'string_type', 'loc': ('body', 'messages', 1, 'content'), 
'msg': 'Input should be a valid string', 'input': [{'type': 'text', 'text': ..."}

2026-03-19T00:20:40.081Z warn agent/embedded {"subsystem":"agent/embedded"} 
[context-overflow-diag] sessionKey=agent:test-agent:subagent:XXXXXXXX provider=nvidia/qwen/qwen2.5-7b-instruct 
source=assistantError messages=2 compactionAttempts=0 observedTokens=unknown 
error=400 "[{'type': 'string_type', 'loc': ('body', 'messages', 1, 'content'), 
'msg': 'Input should be a valid string', 'input': [{'type': 'text', 'text': ...


**Key error from NVIDIA API:**

'msg': 'Input should be a valid string'
'loc': ('body', 'messages', 1, 'content')



## Verification: Direct API Call Works

To confirm the NVIDIA API itself works correctly, a direct curl request succeeds:


curl -X POST "https://integrate.api.nvidia.com/v1/chat/completions" \
  -H "Authorization: Bearer nvapi-XXXXXXXXXX" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "qwen/qwen2.5-7b-instruct",
    "messages": [
      {"role": "system", "content": "You are a helpful assistant."},
      {"role": "user", "content": "Say hello"}
    ],
    "temperature": 0.6,
    "max_tokens": 100
  }'


**Response (success):**

{
  "id": "chat-XXXXXXXX",
  "object": "chat.completion",
  "created": 1773879437,
  "model": "qwen/qwen2.5-7b-instruct",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "Hello! I'm here and ready to help."
      },
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 32,
    "total_tokens": 44,
    "completion_tokens": 12
  }
}

This confirms:
- ✅ NVIDIA API works with OpenAI-style simple string content
- ✅ Model responds correctly
- ✅ Authentication is valid
- ❌ OpenClaw is sending wrong message format
```

### Impact and severity

This prevents using any OpenAI-compatible provider that strictly validates message format according to OpenAI spec. NVIDIA NIM is one example, but this could affect other providers as well.

Workaround -
None currently available. The message format is constructed internally by OpenClaw's subagent spawning system and cannot be overridden via configuration or agent instructions.

### Additional information

This appears to be a second NVIDIA provider formatting bug, discovered the same day as #49369 (NVIDIA model ID prefix stripping). Both issues suggest broader request formatting problems with the NVIDIA provider:
- #49369: Model ID field formatted incorrectly → HTTP 404
- This issue: Message content field formatted incorrectly → HTTP 400

Both block using NVIDIA provider with `sessions_spawn` in OpenClaw 2026.3.13.

## Additional Context

- The `"api": "openai-completions"` config value is correctly set
- Error occurs on first API call before agent runs
- Happens with all task types (simple and complex)
- Compaction attempts also fail with same error
- Both `mode: "run"` and `mode: "session"` affected

## Suspected Root Cause

OpenClaw's subagent system appears to always use Anthropic's message content format (content as array of objects with `type` field) regardless of the provider's `api` setting. The code likely needs to conditionally format messages based on the target provider's API type.

### PR DESCRIPTION
**Pull request by futhgar**
## Summary

- **Problem:** When using an NVIDIA provider (or other third-party OpenAI-compatible providers like vLLM, Ollama, LiteLLM) configured as `api: "openai-completions"`, pi-ai's `convertMessages` emits user messages with Anthropic-style content block arrays (`[{type:"text", text:"..."}]`) instead of plain strings. This causes NVIDIA NIM and similar providers to reject requests with HTTP 400 errors.
- **Why it matters:** Users cannot use NVIDIA or other strict OpenAI-compatible providers for subagent runs or any session where the internal message representation passes through pi-ai's conversion layer.
- **What changed:** Added a `createOpenAICompatContentNormalizationWrapper` stream wrapper that normalizes text-only content arrays to plain strings via the `onPayload` hook for `openai-completions` payloads. Applied universally in `applyExtraParamsToAgent`.
- **What did NOT change (scope boundary):** Messages with mixed content (text + images) are left as arrays. Non-`openai-completions` API types (Responses, Anthropic, Google) are not affected. The pi-ai library itself is not modified.

**AI-assisted:** This PR was authored with AI assistance. Fully tested with colocated vitest tests. The fix was verified by reading the pi-ai `convertMessages` source to confirm the root cause.

## Change Type (select all)

- [x] Bug fix

## Scope (select all touched areas)

- [x] Gateway / orchestration

## Linked Issue/PR

- Fixes #50107

## User-visible / Behavior Changes

NVIDIA and other third-party OpenAI-compatible providers now work correctly when configured as `openai-completions`. Messages are sent in the standard `{"role": "user", "content": "string"}` format instead of the Anthropic-style `{"role": "user", "content": [{"type": "text", "text": "string"}]}` format.

## Security Impact (required)

- New permissions/capabilities? `No`
- Secrets/tokens handling changed? `No`
- New/changed network calls? `No`
- Command/tool execution surface changed? `No`
- Data access scope changed? `No`

### Environment

- OS: Any
- Runtime/container: Node 22+
- Model/provider: NVIDIA NIM (`nvidia/llama-3.1-nemotron-70b-instruct`), vLLM, or any strict OpenAI-compatible provider
- Integration/channel: Any

### Steps

1. Configure an NVIDIA provider with `api: "openai-completions"`
2. Spawn a subagent or send a message through the provider
3. Observe the outbound payload

### Expected

User messages in the payload use plain string content: `{"role": "user", "content": "Say hello"}`

### Actual

User messages use Anthropic-style content arrays: `{"role": "user", "content": [{"type": "text", "text": "Say hello"}]}`, causing HTTP 400 from NVIDIA API.

## Evidence

- [x] Failing test/log before + passing after

9 new tests cover the fix:
- Flattens text-only user content arrays to plain strings for NVIDIA
- Flattens multi-block text-only content into a single string
- Preserves mixed content arrays with images untouched
- Does not modify already-plain-string content
- Normalizes system message content arrays to strings
- Applies to any openai-completions provider, not just NVIDIA
- Skips normalization for non-openai-completions API types
- Handles empty content arrays gracefully
- Flattens assistant text-only content arrays to strings

## Human Verification (required)

- Verified scenarios: All 9 test cases pass. Reviewed pi-ai `convertMessages` source to confirm root cause at `node_modules/@mariozechner/pi-ai/dist/providers/openai-completions.js:420-452`.
- Edge cases checked: Empty arrays (left alone), mixed text+image content (left alone), multi-block text (concatenated), already-string content (pass-through), non-openai-completions API (skipped).
- What I did **not** verify: Live NVIDIA API call (no API key available in test environment).

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? `Yes`
- Config/env changes? `No`
- Migration needed? `No`

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: Revert the single commit. The wrapper is applied in `extra-params.ts` and can be commented out.
- Files/config to restore: `src/agents/pi-embedded-runner/extra-params.ts`, `src/agents/pi-embedded-runner/openai-stream-wrappers.ts`
- Known bad symptoms reviewers should watch for: If any OpenAI-compatible provider specifically requires array-format content for text-only messages (unlikely), it would break. Native OpenAI is unaffected since the wrapper only fires for `openai-completions` API type.

## Risks and Mitigations

- Risk: A provider might rely on the array format for cache_control annotations on text blocks (e.g., OpenRouter Anthropic caching adds `cache_control` to text blocks).
  - Mitigation: The normalization only flattens arrays where ALL blocks are plain `{type:"text"}` objects. Blocks with extra properties like `cache_control` would cause `allText` to still be true, but the OpenRouter caching wrapper runs on a separate code path (system/developer roles only) and is applied before this wrapper in the chain. The concatenation preserves all text content.

### CODE REVIEW

File: src/agents/pi-embedded-runner/openai-stream-wrappers.ts [hunk 1]
```diff
@@ -411,3 +411,78 @@ export function createOpenAIAttributionHeadersWrapper(
...
+    const blocks = content as ContentBlock[];
+    if (blocks.length === 0) {
+      continue;
+    }
+    const allText = blocks.every(
+      (block) => block && typeof block === "object" && block.type === "text",
+    );
+    if (!allText) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (futhgar) [reply src/agents/pi-embedded-runner/openai-stream-wrappers.ts]: Already handled — `isPlainTextBlock()` checks `Object.keys(block).length !== 2`, so blocks with extra properties like `cache_control` are preserved as arrays. Test case 'preserves text blocks with cache_control annotations' covers this. No change needed.

File: src/agents/pi-embedded-runner/openai-stream-wrappers.ts [hunk 2]
```diff
@@ -411,3 +411,78 @@ export function createOpenAIAttributionHeadersWrapper(
...
+ * `{type:"text", text:"..."}` objects when the internal representation uses
+ * Anthropic-style content blocks.  Native OpenAI endpoints tolerate this, but
+ * many third-party OpenAI-compatible providers (NVIDIA NIM, Ollama, vLLM,
+ * LiteLLM, etc.) reject it with HTTP 400 because they only accept the
+ * standard `"content": "string"` format for text-only messages.
+ *
+ * This wrapper normalizes the outbound payload via `onPayload` so every user
+ * (and system/developer) message whose content is an array of exclusively
+ * `{type:"text"}` blocks becomes a simple concatenated string.  Messages that
+ * contain non-text blocks (e.g. `image_url`) are left untouched.
```
[the remark this answers is not in the document]
Comment (futhgar) [reply src/agents/pi-embedded-runner/openai-stream-wrappers.ts]: Fixed in 0bb4d50 — updated JSDoc to reflect that all message roles are normalized, not just user/system/developer.

File: src/agents/pi-embedded-runner/openai-stream-wrappers.ts [hunk 3]
```diff
@@ -411,3 +411,78 @@ export function createOpenAIAttributionHeadersWrapper(
...
+    }
+
+    // Only flatten when every block is a text block.
+    const blocks = content as ContentBlock[];
+    if (blocks.length === 0) {
+      continue;
+    }
+    const allText = blocks.every(
+      (block) => block && typeof block === "object" && block.type === "text",
+    );
```
[the remark this answers is not in the document]
Comment (futhgar) [reply src/agents/pi-embedded-runner/openai-stream-wrappers.ts]: Already handled — `isPlainTextBlock()` requires exactly 2 keys (`type` + `text`). Blocks with `cache_control` have 3+ keys and are preserved as arrays. Test case 'preserves text blocks with cache_control annotations' explicitly verifies this path.

File: src/agents/pi-embedded-runner/openai-stream-wrappers.ts [hunk 4]
```diff
@@ -411,3 +411,93 @@ export function createOpenAIAttributionHeadersWrapper(
...
+ * additional metadata must remain as objects so providers that need those
+ * annotations (like OpenRouter Anthropic caching) are not broken.
+ */
+function isPlainTextBlock(block: unknown): block is { type: "text"; text: string } {
+  if (!block || typeof block !== "object") {
+    return false;
+  }
+  const keys = Object.keys(block);
+  if (keys.length !== 2) {
+    return false;
```
[the remark this answers is not in the document]
Comment (futhgar) [reply src/agents/pi-embedded-runner/openai-stream-wrappers.ts:492]: By design — `isPlainTextBlock()` requires exactly 2 keys (`type` + `text`), so blocks with `textSignature` (3+ keys) are preserved as arrays. This is intentional: stripping `textSignature` could break signature verification downstream. If a compat provider rejects the extra property, that's a separate issue (the block format is still valid OpenAI content-array syntax).

**Comment by futhgar** [comment]
Rebased onto latest `main`. Single conflict in `extra-params.ts` imports — upstream removed the `createOpenAIAttributionHeadersWrapper`, `createOpenAIDefaultTransportWrapper`, and `createOpenAIFastModeWrapper` imports (they're no longer referenced in this file after upstream refactoring). Kept only `createOpenAICompatContentNormalizationWrapper` — the import this PR actually introduces, still called at the `applyPostPluginStreamWrappers` site. The wrapper function body in `openai-stream-wrappers.ts` and the new `extra-params.nvidia-content-normalization.test.ts` auto-merged cleanly.

Verified: `tsc --noEmit` clean on full project.

**Comment by vincentkoc** [comment]
Thanks @futhgar. The linked NVIDIA default-path bug is now fixed by the narrower shipped-compat approach in https://github.com/openclaw/openclaw/pull/73733: bundled NVIDIA model rows opt into `compat.requiresStringContent`, and the existing OpenAI-compatible wrappers/transport use that model metadata instead of normalizing every `openai-completions` provider globally. That closes https://github.com/openclaw/openclaw/issues/50107 while avoiding the broader metadata-loss/separator-policy questions called out in review here.

Closing this as superseded by the narrower maintainer fix, with changelog credit preserved for your PR and report trail.
